### ISSUE TITLE
Shouldn't this line use `every`?

### ISSUE THREAD

**Issue by rixo** [PepsRyuu/nollup#49]
[This line](https://github.com/PepsRyuu/nollup/blob/fbda06386da18638776e50412e25681a760c63d6/lib/plugin-hmr.js#L62), shouldn't it be `every` instead of `some`? We need all impacted branches to independently accept the change, don't we?

**Comment by PepsRyuu** [comment]
Hmm, the original intent of that line is to stop looking for accept callbacks once one is found. Otherwise this could end up accidentally triggering callbacks that are not meant to happen. For example, I might have an accept callback in a CSS file that I want to trigger when I update the CSS file, but I wouldn't want a file in that tree who imports the CSS to trigger.

Not entirely sure if it's as simple as replacing with every.

**Comment by rixo** [comment]
In your example, the accept handler in the depending file that imports the CSS wouldn't be called anyway because the upward bubble [wouldn't propagate](https://github.com/PepsRyuu/nollup/blob/fbda06386da18638776e50412e25681a760c63d6/lib/plugin-hmr.js#L59) beyond the CSS file itself, since it has a accept handler.

But let's say we have 3 modules A, B and C with both B and C importing A, and only B having a accept handler:

~~~
A# <- B*
A# <- C
~~~

(Hashes for changes, asterisks for accept handlers.)

When A changes, the bubble will split between branches B and C. In this situation, my expectation is that the update should not be accepted because the bubble in the C branch will propagate beyond an entry point without encountering an accept handler.

In my opinion, in the following situation the update should be accepted, but both accept handlers should be called:

~~~
A# <- B*
A# <- C*
~~~

One pitfall though is: what should happen in the following case?

~~~
A# <- B <- C*
A# <- C*
~~~

My expectation here is that the update should be accepted _but_ C's accept handler should only be run once.

Maybe this could do it:

~~~js
if (instances[id].invalidate) {
  return !!instances[id].hot._accept
}
~~~

**Comment by PepsRyuu** [comment]
Hmmm, under the assumption that multiple handlers should trigger, I agree with the last two scenarios, but I disagree with the first one, I would expect the handler in B to trigger. 

Has this been tested in other bundlers?

In the meantime, I think I need to write test cases for the current HMR client. This is starting to sound like it could get messy quickly! 😁

**Comment by rixo** [comment]
> Has this been tested in other bundlers?

My description is consistent with Webpack's behaviour. I think. Never too sure when it comes to Webpack...

Webpack doesn't really have accept handlers for self accepted modules though. The callback you pass to `accept` are actually error handlers that are only run when there is an error during the module update. And AFAICT you can't recover from the error, even with a handler. But I digress.

Also, in Webpack self-accepted modules are automatically required again after HMR update (whereas in Nollup -- correct me if I'm wrong -- you need to `require(module.id)` in the accept handler for the update to have any effect at all).

But the overall behaviour matches what I described above.

In the first scenario, Webpack does a full reload.

In the second, the update is accepted.

In the third, the update is accepted and C's "accept handler" is run only once -- necessarily, when you consider that in Webpack that means requiring the whole module...

I've used [this example](https://github.com/webpack/webpack-dev-server/tree/master/examples/cli/hmr) as a playground.

> In the meantime, I think I need to write test cases for the current HMR client.

Ho! Do you happen to have some kind of test around this already? I'm curious to see how you'd tackle the test harness... I ended up with a full scale monstrosity with my HMR test project. Well, project**s**. Hu hu.

**Comment by PepsRyuu** [comment]
Would need to compare this with Parcel, as I've implemented the accept API after their implementation since it seems easier and makes more sense to me.

I've pushed some initial tests for the HMR plugin to the master branch if you're curious as to the approach I've adopted. The accept tests needs far more work still in terms of addressing all of the situations you've outlined. You can find the tests here: https://github.com/PepsRyuu/nollup/blob/master/test/cases/plugin-hmr.js

**Comment by rixo** [comment]
Nice! I think I'll take inspiration from this test approach some day...

I hear you about Webpack hot API. Took me forever to begin to make sense of it.

So! I went ahead and tested the above cases in Parcel. (In fact, I ended up creating [a repository](https://github.com/rixo/test-case-hot-api-webpack-parcel-nollup-rollup) to test the behaviour of each bundler. It implements my 2nd scenario which, I think, is the most tricky. I've added a summary of the results in the README.)

To my great surprise Parcel's results do come different from Webpack.

In the first case, the HMR update is accepted even though the C branch has no accept handler.

In the second, the update is accepted BUT only one of the two possible accept handlers is run. Which one is run depends on the order in which the modules (with accept handlers) are imported.

In the last scenario, no surprise, the update is accepted and the handler is run only once.

OK so I guess this matches Nollup's current behaviour?

I must say I remain very puzzled by this behaviour. Webpack makes more sense to me on this point. It evens seems at odd with Parcel's docs itself:

> You call module.hot.accept with a callback function which is executed when that module or any of its dependencies are updated.

This is not true in my second scenario, since C has a accept handler that is not called although it has A as a dependency and A has been updated.

I'm inclined to think this should be considered a bug in Parcel.

My theory is that the "bug" might go essentially unnoticed in practice because most HMR adapters (e.g. React, Vue...) target Webpack as their first implementation and so they don't rely on accept handlers being called. So this difference in behaviour between Webpack and Parcel / Nollup does not matter. Furthermore, HMR is mostly used to replace UI components (e.g. Svelte, React. Vue...) and, as far as I can tell, these adapters all put an accept handler in every relevant (i.e. component) module. So the difference in regard of rerunning consumer (parent) modules also don't matter for this use case.

For app specific usages of the hot API though (e.g. `clearInterval`...), I think Parcel's behaviour probably causes subtle HMR bugs because it fails to rerun the modules in some of the branches that are affected by a HMR update.

In the example, I consider that module C is broken after the HMR update with Parcel, because it has not been rerun with updated bindings for A. Using this module after the update will produce outdated effects. Further, even though I think it's correct to not rerun the `main` module because the bubble has been accepted before it in all branches, I still consider that `main` is broken after the update with Webpack, Parcel, and Nollup, because it uses outdated bindings (Webpack docs state that bindings are updated automatically when using ES `import`).

Do you you share these conclusions? Or is there something I have overlooked?

Also, my test repo shows that Nollup fails to call dispose handlers for some of the modules it does rerun (module B in the example). I've not investigated but that should probably be an easy fix. Especially with your new tests ;)

And, since we've got our hands in this... While I was experimenting with Parcel, I saw that Nollup's API was mirroring it almost exactly, with the one notable difference that Nollup does not support `module.hot.data` / `module.hot.dispose(data => { ... })`. This is supported with the same API in Webpack. That's a feature I like because it prevents having to mess with outer scope (e.g. `window`) to pass information between two versions of a module, and it's cleanly cut for this job. Do you think this should be added to Nollup?

**Comment by PepsRyuu** [comment]
Apologies for being slow to respond for the past week or so, been incredibly busy.

That's some amazing research, I love the comparison table, it really helps to understand the differences in behaviours. From the looks of it, I think the hot plugin has the behaviour I would agree with, and I do think it's a bug in Parcel and possibly in Webpack. It's very likely as you say that due to how HMR is used, that these specifics went unnoticed. 

Here's what I propose to be done:

* Update ```dispose``` so that each time we climb the tree, we run the method for those modules, until we hit an accept handler.

* Climb up each parent of an updated module, until we hit an ```accept``` callback for that branch. 

* Add ```module.hot.data```. 

A big thing to note here is live-bindings. I'm purposely ignoring implementing live-bindings, due to the performance penalty of scanning the code and the inaccuracies that can arise from that. It's a really underused feature of ESM, and I'm not sure if it's worthwhile implementing.

With the new test suite, I think all of these can be implemented and tested fairly quickly now. Let me see what I can come up with.

**Comment by PepsRyuu** [comment]
Added new test cases with an updated approach to dispose and accept handlers. Now runs multiple accept and dispose handlers: https://github.com/PepsRyuu/nollup/tree/HMRAlignment

Need to do some research into ```module.hot.data```. Should be done soon.

**Comment by PepsRyuu** [comment]
Added implementation of ```module.hot.data``` with tests.

**Comment by rixo** [comment]
Great!

I'm essentially OK with your proposed plan above.

I've tried your new branch, and quickly reviewed the changed code. Now, it completely matches my expectations, except for the rebinding thing.

Some remarks...

**module.hot.data**

Regarding `module.hot.data`, I realized the 4 solutions (~bundlers) all slightly different behaviors. On this point, I'm gonna align `rollup-plugin-hot` with what Webpack does:

- module.hot.data is `undefined` on initial module load (with Parcel, it is `null`; with Nollup, it is an empty object)

- module.hot.data defaults to {} after a HMR update, _even if the module has no dispose handlers_ (mine was still `undefined`)

**What do you do with "unaccepted" updates?**

i.e. if an update bubbles up beyond an entry point without encountering an accept handler, what do you do?

My observation is that Nollup is ignoring the update. Correct?

All other solutions order a full reload in this case. Full reloads can be prevented with an option in Webpack (if I remember correctly), and my plugin. I don't know for Parcel.

As a HMR user, my expectation is a full reload any time HMR loses sync. That is, HMR works like LiveReload, except in the cases where it can do better. But it shouldn't do less than LiveReload.

An option to disable full reloads is still useful because, in some projects HMR, keeps failing all the time, and the full reload policy can prevent the user from seeing useful errors.

**Timing of dispose / accept**

Oh! There _is_ a "dispose phase". This is something I had overlooked until now, not having implemented statuses for now. Not sure I will implement them, but disposing everything before running new code does seem like something I should fix.

**Comment by rixo** [comment]
I don't think the last problem is related to live bindings. It's just that you're rebinding non reloaded module with stale values. Just [bypassing the cache](https://github.com/rixo/nollup/commit/f4b3efe1054fe38347799fa138385dd74b1f06ae) fixes the "Rebind main" issue in my test scenario, making Nollup all good on the tested criteria. (I'm not saying my hack is the right implementation.)

**Comment by rixo** [comment]
Added a [commit](https://github.com/rixo/nollup/commit/bf8e078881456240aa3d2c69a002d9339700c008#diff-442ff38fd177df1d0884a3e226a940aaR86) to my branch because accept handlers can get reset by running other accept handlers with my previous change. Not enough cache this time.

With this, Nollup gets all the marks in my comparison table.

**Comment by PepsRyuu** [comment]
Updated ```module.hot.data``` to change defaults and to become an object after HMR update.

Regarding unaccepted, correct, if there's no accept or dispose handlers, nothing will happen. Personally I'm not a fan of when pages auto-refreshing unless it's something that's explicitly opted into. I'm okay with having a ```reload` that defaults to false, but if set to true, will refresh if no accept handlers were found.

Do you have a test scenario for the live binding? Want to make sure I understand the exact situation you're referring to.

**Comment by rixo** [comment]
> I'm okay with having a reload that defaults to false

That's the kind of privilege you get for coding your own bundler. You pick the defaults.

But you should at least have something in the console to indicate that what's on screen is out of sync, no? Pretty visible if possible.

> if set to true, will refresh if no accept handlers were found

Not sure that's enough. There are cases of error during updates that also warrant a full reload (IMHO). Don't do it for me now, I'd need to clarify my demand first...

> Do you have a test scenario for the live binding?

Well, as a matter of fact I've just created [one](https://github.com/rixo/repro-systemjs) lol. But it's not for Nollup, and it's unrelated. While digging this, I've found a bug with live bindings in SystemJS...

But the [test repo](https://github.com/rixo/test-case-hot-api-webpack-parcel-nollup-rollup) we've been using until now will do. I don't believe the problem I'm talking about is related to live bindings. Just normal bindings.

Here's the problem. With your HMRAlignment branch, I now get the following logs:

~~~
b.js: a
c.js: a
main.js
main.js: b(a) c(a)
~~~

After change to `a.js`:

~~~
dispose a
dispose b.js
dispose c.js
accept b.js a
a.js := a UPDATED
b.js: a UPDATED
accept c.js a
c.js: a UPDATED
main.js: b(a) c(a)
~~~

As far as I'm concerned, this is mostly correct, exept for the very last line:

~~~
main.js: b(a) c(a)
~~~

Should be:

~~~
main.js: b(a UPDATED) c(a UPDATED)
~~~

With the changes on [my branch](https://github.com/rixo/nollup/tree/HMRAlignment), this is fixed.

I've put a debugger in there. I've seen that the binder function of `main.js` was correctly being called again after the update. But it is reading from `_i0` and such, that are [never updated](https://github.com/PepsRyuu/nollup/blob/f87fae8cf89e02659e3176f8310b16ea6f7a184b/lib/index.js#L177), and so it rebinds the same stale value. My change just require the deps directly in the binder function, getting fresh ones each time. (This in turn broke your new `hmrAcceptCallback` because modules are losing their accept handler when other accepted modules require them, hence my second change.)

**Comment by PepsRyuu** [comment]
I think I'm going to need to think about the user experience of HMR more. Not a fan of throwing messages that the UI is out of sync, as it could potentially interfere with the debugging experience. Will get back to this point.

Currently very interested in the binding. I think I understand the binding problem now. Yeah I don't think it's exactly the live-binding problem I was thinking of, this is more of a "re-export" problem, which is a subset of the live-binding problem I have not previously considered. I can certainly have a look at this in more detail. Your solution seems like it'll be fine but just need to try a few tests with it.

**Comment by PepsRyuu** [comment]
Updated with "re-export" problem fixed.

**Comment by rixo** [comment]
Huge! Is this pushed?

**Comment by PepsRyuu** [comment]
To master no, still on branch, but you can see the PR ready to go.

**Comment by rixo** [comment]
Ah yes! Got tripped by the commit date (still wondering).

Almost there!

~~~
a.js := a
b.js: a
c.js: a
main.js
main.js: b(a) c(a)
dispose a
dispose b.js
dispose c.js
accept b.js a
a.js := ax
b.js: ax                   <-- missing instantiation of c
main.js: b(ax) undefined          <-- should be b(ax) c(ax)
~~~

With `require` now being called during other modules' binding, a module can be replaced before it is its turn in the loop in `acceptCallback`. Here's how I fixed it, that should explain it more clearly: https://github.com/rixo/nollup/commit/bf8e078881456240aa3d2c69a002d9339700c008

I fail to appreciate if my solution is the best approach. My first intuition was to try to maintain the module reference's identity in `instances` and store the module reference in `_i0`. Then, in the binder, do `local = _i0.exports.default; local2 = _i0.exports.named`. `exports` changes, modules (`_i0`) stays. 

That would not keep the require order as it is now. But in turn, I'm not entirely sure that the module would be resolved at the time the binder is called with the current flow. I think you're in a better position than me to determine the best solution here.

**Comment by PepsRyuu** [comment]
Good catch, adding a new test case to address this specific scenario. Let me think about this one.

**Comment by PepsRyuu** [comment]
Pushed what is hopefully the final commit for this. 

I've changed the implementation significantly. Instead of blindly going up the dependency tree looking for dispose and accept handlers, the update function will now track individual branches. If there's an accept handler on that branch, it will trigger that accept and all of the dispose handlers on that branch. If there's no accept on that branch, then the dispose handlers will not trigger.

There's now also a separation of module creation and getting exports. It's a bit of a weird thing to explain so here's what happened:

```
function (create_bindings, define_module, export_binding) {
    // Create placeholder variables for storing exports.
    var i0;

    // Create placeholders for imports
    var MyImport;

    define_module(function () {
        // This binder function is called every time a child dependency
        // has called export_binding. We reimport those bindings and 
        // reassign the variables.
        MyImport = i0().MyImport;
    }, function () {
        // Module code
        export_binding('default', 'Hello World');
    });

    i0 = () => create_bindings(0)
}
```

How it roughly works, when you call ```require()```, it will first call ```create_bindings```. This function will create an instance of the module with an empty exports object. It will then run the above function code. Nothing will happen initially, defining a module only stores the binder and module callbacks. The last thing ```create_bindings``` does is call the binder function. The binder function calls ```create_bindings``` again. As a result, it builds a dependency tree, instantiating empty export objects for the entire bundle. 

Once all of the empty export objects have been created, ```require()``` goes into the second phase, ```resolve_module```. This function will execute the module code, and will do this recursively for all of the dependencies. ```export_binding``` will eventually be called inside these modules. This function will update the exports for the module, and it will call all of the parent modules, and recall their binders so that you're working with the latest value.

This separation was necessary in order to support circular dependencies, as circular dependencies only works with ESM when you have live-bindings. In the current implementation, live-bindings only apply on initial module execution. It theoretically can work as well for deferred binding updates by re-running ```export_binding```. 

This works well, until you hit multiple branches. What was happening in the multi-branch case was the following:

```
A imports B + C with accept handlers who both import D
A <-- B <-- D
A <-- C <-- D
```

* D updates.
* D, B, C marked for disposal and invalidated.
* B runs accept handler, requires itself.
* B requires D
* Requiring calls both ```create_bindings``` and ```resolve_module``` which calls ```export_binding```.
* B and C binder functions run because of D. 
* B runs its module code, calling its own ```export_binding```.
* A binder function runs because of B, calls ```create_bindings``` for B and C. 
* ```create_bindings``` will wipe out a module if it is invalidated, C has been marked as invalidated.
* A receives the new exports for B, and receives empty exports for C.
* C no longer has an accept handler because it has been wiped out. Can't require itself to fix things.

The accept handler workaround with cloning the accept handlers before running them isn't ideal, because it still ends up with C incorrectly wiping itself out. It ends up with a satisfactory result only because C eventually requires itself.

To solve this, it was necessary to separate the creation of modules from getting their exports:

```
function (create_module, get_exports, define_module, export_binding) {
    // Create placeholder variables for storing exports.
    var i0;

    // Create placeholders for imports
    var MyImport;

    define_module(function () {
        // This binder function is called every time a child dependency
        // has called export_binding. We reimport those bindings and 
        // reassign the variables.
        MyImport = i0().MyImport;
    }, function () {
        // Module code
        export_binding('default', 'Hello World');
    });

    i0 = create_module(0) && () => get_exports(0)
}
```

Here's how this works now: 

* D updates.
* D, B, C marked for disposal and invalidated.
* B runs accept handler, requires itself.
* B requires D. Both are marked as invalidated, which means calling ```create_module``` and ```resolve_module``` which ends up calling ```export_binding```.
* B and C binder functions run because of D exporting. 
* B runs its module code, calling ```export_binding```.
* A binder function runs because of B, calls ```get_exports``` for B and C. 
* A receives new exports for B, and receives old exports for C.
* C still has its accept handler, this now runs.
* C runs require, it's marked as invalidated, so it calls ```create_module``` and ```resolve_module```, and then ```export_binding```.
* A binder function runs again, calls ```get_exports``` for B and C.
* A receives the previous exports for B and the new exports for C.

I'm much happier with this solution, and have updated the test cases with more scenarios. I think that should about wrap this up as far as I can tell unless there's another oversight I'm missing.

**Comment by rixo** [comment]
Man, you're the man!

With this, Nollup's all clear in the test case! I've updated [the results](https://github.com/rixo/test-case-hot-api-webpack-parcel-nollup-rollup#results) accordingly :) :tada:

The new version is failing Svelte HMR e2e suite with a binding problem though.

The issue can be reproduced with [svelte-template-hot](https://github.com/rixo/svelte-template-hot):

~~~
git clone git@github.com:rixo/svelte-template-hot.git
cd svelte-template-hot
yarn
yarn add -D PepsRyuu/nollup#HMRAlignment

yarn dev:nollup
~~~

Open the page at http://localhost:5000: you're met with an error:

~~~
Uncaught TypeError: makeApplyHmr is not a function
~~~

In the relevant source code, `makeApplyHmr` is an import:

~~~js
import { makeApplyHmr } from 'svelte-hmr/runtime';

export const applyHmr = makeApplyHmr(args =>
	Object.assign({}, args, {
		hot: args.m.hot,
	})
);
~~~

It works with Rollup (`yarn dev:rollup`) or previous versions of Nollup...

I'll try to see if I can get more info.

**Comment by PepsRyuu** [comment]
Seems to be something to do with ```export { } from```. That was a problem when making the changes to the binding mechanism. Probably a small oversight somewhere.

**Comment by PepsRyuu** [comment]
Working now. :)

On a side note, the template doesn't work with Windows out of the box (environment variables not being applied and the path to the Nollup compatibility layer not replacing "\" with "/").

**Comment by rixo** [comment]
Everything seems good for me now. All tests passing. Congratulations! And thank you!

Thanks also for the issue about the template.

I think we're done here. I suppose the issue should be closed when the PR will be merged.

**Comment by PepsRyuu** [comment]
Excellent! Will be merged once I've finished with some other big changes. :)

**Comment by PepsRyuu** [comment]
Released in ```0.9.0```
